File: Ahorn/entities/SaloonChandelier.jl
module SaloonChandelier

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/SaloonChandelier" SaloonChandelier(x::Integer, y::Integer, canFall::Bool=true, chainLength::Integer=80, isHazard::Bool=true, swingAngle::Number=0.4, swingPeriod::Number=3.0)

const placements = Ahorn.PlacementDict(
    "normal" => Ahorn.EntityPlacement(SaloonChandelier)
)

function Ahorn.selection(entity::SaloonChandelier)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SaloonChandelier, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/saloonChandelier.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local saloonChandelier = {}

saloonChandelier.name = "MaggyHelper/SaloonChandelier"
saloonChandelier.depth = -100
saloonChandelier.fieldInformation = {
    swingPeriod = {
        minimumValue = 0
    },
    swingAngle = {
        minimumValue = 0
    },
    chainLength = {
        minimumValue = 0
    }
}
saloonChandelier.placements = {
    {
        name = "normal",
        data = {
            swingPeriod = 3.0,
            swingAngle = 0.4,
            chainLength = 80,
            canFall = true,
            isHazard = true
        }
    }
}

function saloonChandelier.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/saloon_chandelier/swinging", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 0.0)
    return sprite
end

function saloonChandelier.selection(room, entity)
    return utils.rectangle(entity.x - 20, entity.y, 40, 24)
end

return saloonChandelier
